lib/image-stream/parse-partitions.js

```diff
@@ -0,0 +1,166 @@
+  // For example, ISOs will usually have a block size of 4096,
+  // but raw images a block size of 512 bytes
+  while (blockSize <= MAX_BLOCK_SIZE) {
+    try {
+      gpt = GPT.parse(buffer.slice(blockSize));
+    } catch (error) {
+      // Ignore error
+    }
+    if (gpt) {
+      return gpt;
+    }
+    blockSize += blockSize;
+  }
+
+  return gpt;
+
+};
+
+/**
+ * Attempt to parse the MBR & GPT from a given buffer
+ * @private
+ * @param {Object} image - Image metadata
+ * @param {Buffer} buffer - Buffer
+ * @returns {undefined}
+ */
+const parsePartitions = (image, buffer) => {
+
+  let mbr = null;
+  let gpt = null;
+
+  try {
+    mbr = MBR.parse(buffer);
+    image.hasMBR = true;
+    if (mbr.getEFIPart()) {
+      image.hasGPT = true;
+      gpt = parseGPT(buffer);
+    }
+  } catch (error) {
+    // Ignore error
+  }
+
+  // As MBR and GPT partition entries have a different structure,
+  // we normalize them here to make them easier to deal with and
+  // avoid clutter in what's sent to analytics
+  if (image.hasGPT) {
+    image.partitions = _.map(gpt.partitions, (partition) => {
+      return {
+        type: partition.type.toString(),
+        id: partition.guid.toString(),
+        name: partition.name,
+        start: partition.firstLBA,
+        end: partition.lastLBA,
+        extended: false
+      };
+    });
+  } else if (image.hasMBR) {
+    image.partitions = _.map(mbr.partitions, (partition) => {
+      return {
+        type: partition.type,
+        id: null,
+        name: null,
+        start: partition.firstLBA,
+        end: partition.lastLBA,
+        extended: partition.extended
+      };
+    });
+  }
+
+};
+
+/**
+ * Attempt to read the MBR and GPT from an imagestream
+ * NOTE: This operation will consume the first `MAX_BYTES`
+ * of the stream and then destroy the stream.
+ * @param {Object} image - image metadata
+ * @returns {Promise}
+ */
+module.exports = (image) => {
+  return new Bluebird((resolve, reject) => {
+
+    const chunks = [];
+    let length = INITIAL_LENGTH;
+    let destroyed = false;
+
+    image.hasMBR = false;
+    image.hasGPT = false;
+
+    let stream = image.stream.pipe(image.transform);
+
+    stream.on('error', reject);
+
+    // We need to use the "old" flowing mode here,
+    // as some dependencies don't implement the "readable"
+    // mode properly (i.e. bzip2)
+    stream.on('data', (chunk) => {
+      chunks.push(chunk);
+      length += chunk.length;
+
+      // Once we've read enough bytes, terminate the stream
+      if (length >= MAX_BYTES && !destroyed) {
+
+        // Prefer close() over destroy(), as some streams
+        // from dependencies exhibit quirky behavior when destroyed
+        if (image.stream.close) {
+          image.stream.close();
+        } else {
+          image.stream.destroy();
+        }
+
+        // Remove references to stream to allow them being GCed
+        image.stream = null;
+        image.transform = null;
```

Comment:
Yeah, I just thought it might be more explicit to keep everything localised - if they're being set to null anyway, might make sense to delete them here and then remove the `.omitBy` from the other place?